voltdrop_direct/index.js

```diff
@@ -0,0 +1,448 @@
+ * @property {string[]} warnings - A list of warning messages that do not prevent the driver from encoding the downlink object
+ */
+
+/**
+ * Downlink encode
+ * @param {Object} input - An object provided by the IoT Flow framework
+ * @param {Object} input.data - The higher-level object representing your downlink
+ * @returns {EncodedDownlink} The encoded object
+ */
+function encodeDownlink(input) {
+  let result = {
+    bytes: [],
+    errors: [],
+    warnings: [],
+  };
+
+  let definedDownlinkVars = 0;
+  if (typeof input.data.transmitIntervalSeconds !== "undefined") {
+    definedDownlinkVars += 1;
+  }
+  if (typeof input.data.packetTransmitSchedule !== "undefined") {
+    definedDownlinkVars += 1;
+  }
+  if (typeof input.data.factoryReset !== "undefined") {
+    definedDownlinkVars += 1;
+  }
+  if (typeof input.data.softReset !== "undefined") {
+    definedDownlinkVars += 1;
+  }
+
+  if (definedDownlinkVars > 1) {
+    result.errors.push("Invalid downlink: More than one downlink type defined");
+    delete result.bytes;
+    return result;
+  }
+
+  if (typeof input.data.transmitIntervalSeconds !== "undefined") {
+    if (input.data.transmitIntervalSeconds < 60) {
+      result.errors.push(
+        "Invalid downlink: transmit interval cannot be less than 1 min"
+      );
+      delete result.bytes;
+      return result;
+    }
+    if (input.data.transmitIntervalSeconds > 1800) {
+      result.errors.push(
+        "Invalid downlink: transmit interval cannot be greater than 30 min"
+      );
+      delete result.bytes;
+      return result;
+    }
+    let downlink = Buffer.alloc(6);
+    downlink.writeUInt16BE(0x0031, 0);
+    downlink.writeUInt32BE(input.data.transmitIntervalSeconds, 2);
+    result.bytes = Array.from(new Uint8Array(downlink.buffer));
+    result.fPort = 3;
+    return result;
+  }
+
+  if (typeof input.data.packetTransmitSchedule !== "undefined") {
+    if (input.data.packetTransmitSchedule.length < 1) {
+      result.errors.push(
+        "Invalid downlink: Packet transmit schedule must contain at least one element"
+      );
+      delete result.bytes;
+      return result;
+    }
+    if (input.data.packetTransmitSchedule.length > 60) {
+      result.errors.push(
+        "Invalid downlink: Packet transmit schedule cannot be longer than 60 elements"
+      );
+      delete result.bytes;
+      return result;
+    }
+
+    const validIDList = [0, 40, 41, 42, 43, 44, 45];
+    let scheduleValid = false;
+    let downlink = Buffer.alloc(input.data.packetTransmitSchedule.length + 3);
+    downlink.writeUInt16BE(0x0030, 0);
+    downlink.writeUInt8(input.data.packetTransmitSchedule.length, 2);
+    for (let index = 0; index < input.data.packetTransmitSchedule.length; index++) {
+      let id = input.data.packetTransmitSchedule[index];
+      if (!validIDList.includes(id)) {
+        result.warnings.push("Invalid packet ID replaced by gap in transmit schedule");
```

Comment:
Maybe it would be useful to the user to have the index of the packet ID that is replaced?

Reply:
yeah, that's a good idea.